contracts/GenericLender/GenericAave.sol

```diff
@@ -74,24 +111,118 @@ contract GenericAave is GenericLenderBase {
+    function _incentivesRate(uint256 totalLiquidity) public view returns (uint256) {
+        // only returns != 0 if the incentives are in place at the moment. 
+        // it will fail if the isIncentivised is set to true but there is no incentives
+        if(isIncentivised && block.timestamp < _incentivesController().getDistributionEnd()) {
+            uint256 _emissionsPerSecond;
+            (, _emissionsPerSecond, ) = _incentivesController().getAssetData(address(aToken));
+            if(_emissionsPerSecond > 0) {
+                uint256 emissionsInWant = _AAVEtoWant(_emissionsPerSecond); // amount of emissions in want
+
+                uint256 incentivesRate = emissionsInWant.mul(SECONDS_IN_YEAR).mul(1e18).div(totalLiquidity); // APRs are in 1e18 
+
+                return incentivesRate.mul(9_500).div(10_000); // 95% of estimated APR to avoid overestimations
+            }
+        }
+        return 0;
     }
 
-    //emergency withdraw. sends balance plus amount to governance
-    function emergencyWithdraw(uint256 amount) external override management {
-        _lendingPool().withdraw(address(want), amount, address(this));
+    function aprAfterDeposit(uint256 extraAmount) external view override returns (uint256) {
+        // i need to calculate new supplyRate after Deposit (when deposit has not been done yet)
+        DataTypes.ReserveData memory reserveData = _lendingPool().getReserveData(address(want));
 
-        want.safeTransfer(vault.governance(), want.balanceOf(address(this)));
+        (uint256 availableLiquidity, uint256 totalStableDebt, uint256 totalVariableDebt, , , , uint256 averageStableBorrowRate, , , ) =
+            protocolDataProvider.getReserveData(address(want));
+
+        uint256 newLiquidity = availableLiquidity.add(extraAmount);
+
+        (, , , , uint256 reserveFactor, , , , , ) = protocolDataProvider.getReserveConfigurationData(address(want));
+
+        (uint256 newLiquidityRate, , ) =
+            IReserveInterestRateStrategy(reserveData.interestRateStrategyAddress).calculateInterestRates(
+                address(want),
+                newLiquidity,
+                totalStableDebt,
+                totalVariableDebt,
+                averageStableBorrowRate,
+                reserveFactor
+            );
+
+        uint256 incentivesRate = _incentivesRate(newLiquidity.add(totalStableDebt).add(totalVariableDebt)); // total supplied liquidity in Aave v2
+        return newLiquidityRate.div(1e9).add(incentivesRate); // divided by 1e9 to go from Ray to Wad
+    }
+
+    function hasAssets() external view override returns (bool) {
+        return aToken.balanceOf(address(this)) > 0;
+    }
+
+    // Only for incentivised aTokens
+    // this is a manual trigger to claim rewards once each 10 days
+    // only callable if the token is incentivised by Aave Governance (_checkCooldown returns true)
+    function harvest() external {
+        if(_checkCooldown()) {
```

Comment:
should this be a require instead? this method doesn't seem to do anything if `_checkCoolDown()` doesnt return true

Reply:
yeah

I used an if to avoid spending too much gas if called when conditions were not met but changing this to require sounds good too